### PR TITLE
Feature/change item mapping

### PR DESCRIPTION
**Pull request by millerts**

### CODE REVIEW

File: src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java:34]: I had to add a discriminator column to the table.  I couldn't find a way to have JPA do if stim use Stimulus class otherwise use item class.

So the classification table now has:
* stim 
* item

File: src/main/java/org/opentestsystem/ap/imrt/common/model/ImrtItem.java
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/common/model/ImrtItem.java]: I wanted initial comments before I make another change to the db to remove the composite primary key on the stimlink table.

**Review by millerts — COMMENTED** [review]
I need to add a few more tests around this area but wanted to get initial thoughts

File: src/main/java/org/opentestsystem/ap/imrt/common/repository/BaseItemRepository.java
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/common/repository/BaseItemRepository.java]: The CrudRepository probably isn't appropriate here and I'll need to do something like [this](https://stackoverflow.com/questions/11107136/creating-a-read-only-repository-with-springdata?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa) before I merge

File: src/test/java/org/opentestsystem/ap/imrt/common/repository/ImrtItemRepositoryTest.java
Comment (millerts) [inline src/test/java/org/opentestsystem/ap/imrt/common/repository/ImrtItemRepositoryTest.java]: This will be cleaned up before merging.  I was just testing things.

File: src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java
Comment (malaffoon) [inline src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java]: I don't grok this but i don't really have to ...

File: src/main/java/org/opentestsystem/ap/imrt/common/model/ImrtItem.java
Comment (malaffoon) [inline src/main/java/org/opentestsystem/ap/imrt/common/model/ImrtItem.java]: Logically this is a zero-to-one, right? You couldn't wrangle JPA to do that for you? i.e. make this `private StimLink stimLink`?

File: src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java:42]: Name it however you like, and I stink at naming, but in RDW we have "id" as the generated primary key and "natural_id" as the client identifier.

File: src/main/java/org/opentestsystem/ap/imrt/common/model/Stimulus.java
Comment (malaffoon) [inline src/main/java/org/opentestsystem/ap/imrt/common/model/Stimulus.java]: I realize you are using a common data model for this but it seems odd that these are `StimLink`'s. Because they are really `ItemLink`'s aren't they?

**Comment by malaffoon** [comment]
This seems about right. (I'm suppressing all instincts to complain about javadocs and stuff)

File: src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java:125]: nit: do you really want your setters to be public and your models to be mutable?

File: src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java
Comment (wtritch) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java:34]: Seems reasonable to me.

File: src/main/java/org/opentestsystem/ap/imrt/common/model/ImrtItem.java
Comment (wtritch) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/ImrtItem.java]: I'm a big fan of auto-incrementing single keys.  Being able to reference something by unique id helps a lot with API design.

File: src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java
Comment (millerts) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java]: Yeah, so we have the item json that is created by IAT and IMRT's item.  This is the Java representation of the item.json that is created by IAT.  The naming is going to be a bit tricky within IMRT because we have our "item" and IAT's item but at a base level they are the same.

File: src/main/java/org/opentestsystem/ap/imrt/common/model/ImrtItem.java
Comment (millerts) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/ImrtItem.java]: I need to redo stimlink to not have the composite key.  Any time I try to create a OneToOne JPA gets angry because Stimlink has the composite key within the code.

File: src/main/java/org/opentestsystem/ap/imrt/common/repository/BaseItemRepository.java
Comment (wtritch) [reply src/main/java/org/opentestsystem/ap/imrt/common/repository/BaseItemRepository.java]: A quick test of this polymorphic repo might be a method name like `findByStimLinkStimLinkKeyItemStimKey(int stimKey);`
[per the docs](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.query-methods.query-property-expressions)
Which should hopefully be parsed into "x.stimLink.stimLinkKey.itemStim.key"  And would confirm you could query by a sub-class' properties from the abstract repo.
Side-note after digging through the concept of a stimLink/stimLinkKey I'm a little confused.  Why wouldn't an ImrtItem just have a "stimulus" property that contains a (nullable) Stimulus?

File: src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java
Comment (millerts) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java:42]: Good thought.  It is on our radar to discuss naming in general once Sue is back.

File: src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java
Comment (millerts) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java:125]: How did you manage updates to items within JPA without the ability to modify the object directly?  It seems like JPA (hibernate) doesn't like immutablity by default

Comment (jeffjohnson9046) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java:125]: I'm curious about that myself.  We had that problem in TDS - we added public setters so that JPA could hydrate the objects.  We put all the setters at the bottom of the code file (which goes against normal convention, I know) with a big comment that said "DO NOT USE THESE - they're only for hibernate" (paraphrasing).  If there's another way to do it & maintain immutability (like tell JPA/Hibernate to use a builder?) I'd love to know what it is.

File: src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java
Comment (jeffjohnson9046) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java]: Makes sense to me - this is just IMRT's representation of an Item.

File: src/main/java/org/opentestsystem/ap/imrt/common/model/ImrtItem.java
Comment (jeffjohnson9046) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/ImrtItem.java]: I agree w/@wtritch.  Single field primary keys are a lot easier to deal with.  I understand the data lends itself well to a composite key, but I think that can be enforced with a unique index between the two fields.  We could add an auto-incrementing `int` field (or UUID or whatever) to the `stim_link` table; I've done that in the past for "join" tables and had some success with it.

File: src/main/java/org/opentestsystem/ap/imrt/common/model/Stimulus.java
Comment (jeffjohnson9046) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/Stimulus.java]: On a more general level I agree that it is an item link - it links one item to another.  In this case I think the intent here is to link a Stimulus to an Item (which is the only item association I'm aware of in the SB ecosystem).  From that perspective, I think it makes sense to call it a stim link (although I'd probably spell it out to be `StimulusLink`).

File: src/main/java/org/opentestsystem/ap/imrt/common/repository/BaseItemRepository.java
Comment (jeffjohnson9046) [reply src/main/java/org/opentestsystem/ap/imrt/common/repository/BaseItemRepository.java]: @wtritch 's side note is interesting.  I think we've expressed that in the database well enough, but I think we might have a simpler and more expressive object model if `ImrtItem` had an `Optional<Stimulus>` property.

**Review by jeffjohnson9046 — APPROVED** [review]
Some thoughts/musings, but I don't think there's anything that should hold this up getting merged.

I like @wtritch 's idea of adding an `Optional<Stimulus>` property to the `ImrtItem`, but my concern is that might be a bigger change than we want to make right now.  Not being intimately familiar with the code, I'll defer to others on whether to make that change.

File: src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java
Comment (malaffoon) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java:125]: I don't think Will is suggesting true immutability. Just "hide" the setters so only JPA can get at them. It may have changed but JPA used to be able to call private setters.

Comment (wtritch) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java:125]: We used builders with a copy() method to create a new instance with updated values (if code was doing the modification.)  I think JPA can use a package-local setter just fine, can't it?

File: src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java
Comment (malaffoon) [inline src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java]: Hmm, didn't notice this the first time. You don't want nulls? But you don't have a check in the getter/setter to enforce it. btw, there is a an empty string constant in a few libraries you could use.

File: src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java
Comment (millerts) [reply src/main/java/org/opentestsystem/ap/imrt/common/model/BaseItem.java:125]: Understood, and yes Will.

**Pull request merged**
